drivers/ukconsole/ns16550/ns16550.c

```diff
@@ -253,143 +337,140 @@ static int early_init(struct ukplat_bootinfo *bi)
 
 	return 0;
 }
-#endif /* !CONFIG_LIBUKTTY_NS16550_EARLY_CONSOLE */
+#endif /* !CONFIG_LIBUKCONSOLE_NS16550_EARLY_CONSOLE */
 
-static int init(struct uk_init_ctx *ictx __unused)
+#if CONFIG_LIBUKALLOC
+static int fdt_get_device(struct ns16550_device *dev, const void *dtb,
+			  int *offset)
 {
-	struct ukplat_bootinfo *bi;
-	const __u64 *regs;
-	const void *dtb;
-	int offset, len;
-	__u64 reg_base;
-	__u64 reg_size;
+	__u64 reg_base, reg_size;
 	int rc;
 
-	bi = ukplat_bootinfo_get();
-	UK_ASSERT(bi);
-
-	dtb = (void *)bi->dtb;
+	UK_ASSERT(dev);
 	UK_ASSERT(dtb);
+	UK_ASSERT(offset);
 
-	uk_pr_debug("Probing ns16550\n");
+	*offset = fdt_node_offset_by_compatible_list(dtb, *offset,
+						     fdt_compatible);
+	if (unlikely(*offset < 0))
+		return -ENOENT;
 
-	/* Pick the 1st device in the dtb */
-	offset = fdt_node_offset_by_compatible_list((void *)bi->dtb, -1,
-						    fdt_compatible);
-	if (unlikely((offset < 0))) {
-		uk_pr_err("ns16550 not found in fdt\n");
-		return -ENODEV;
-	}
-
-	rc = fdt_get_address((void *)bi->dtb, offset, 0, &reg_base, &reg_size);
+	rc = fdt_get_address(dtb, *offset, 0, &reg_base, &reg_size);
 	if (unlikely(rc)) {
 		uk_pr_err("Could not parse fdt node\n");
-		return rc;
+		return -EINVAL;
 	}
 
-#if CONFIG_PAGING
-	/* Map device region */
-	ns16550_uart_base = uk_bus_pf_devmap(reg_base, reg_size);
-	if (unlikely(PTRISERR(ns16550_uart_base))) {
-		uk_pr_err("Could not map ns16550\n");
-		return PTR2ERR(ns16550_uart_base);
-	}
-#else /* !CONFIG_PAGING */
-	ns16550_uart_base = reg_base;
-#endif /* !CONFIG_PAGING */
+	uk_pr_debug("ns16550 @ 0x%lx - 0x%lx\n", reg_base, reg_base + reg_size);
 
-	regs = fdt_getprop(dtb, offset, "reg-shift", &len);
-	if (regs)
-		ns16550_reg_shift = fdt32_to_cpu(regs[0]);
+	uk_console_init(&dev->dev, "NS16550", &ns16550_ops, 0);
+	dev->base = reg_base;
+	dev->size = reg_size;
 
-	regs = fdt_getprop(dtb, offset, "reg-io-width", &len);
-	if (regs)
-		ns16550_reg_width = fdt32_to_cpu(regs[0]);
+	return 0;
+}
 
-	uk_pr_debug("ns16550 @ 0x%lx - 0x%lx\n", reg_base, reg_base + reg_size);
+static int register_device(struct ns16550_device *dev, struct uk_alloc *a)
+{
+	struct ns16550_device *console_dev;
 
-	rc = init_ns16550();
-	if (unlikely(rc)) {
-		uk_pr_err("Could not initialize ns16550\n");
-		return rc;
+	UK_ASSERT(dev);
+	UK_ASSERT(a);
+
+	console_dev = uk_malloc(a, sizeof(*console_dev));
+	if (unlikely(!console_dev)) {
+		uk_pr_err("Could not allocate ns16550 device\n");
+		return -ENOMEM;
 	}
 
-	ns16550_uart_initialized = 1;
-	uk_pr_info("tty: ns16550\n");
+	console_dev->dev = dev->dev;
+	console_dev->base = dev->base;
+	console_dev->size = dev->size;
+
+	uk_console_register(&console_dev->dev);
+	uk_pr_info("console: ns16550 (%p)\n", &console_dev->base);
```

Comment:
Maybe we should add also here the interface number since we should got one after register.
```
uk_pr_info("console %"__PRIu16": ns16550 (%p)\n", console_dev->id, &console_dev->base);
```

Reply:
👍🏻 as we discussed offline, we adopted to `con%" __PRIu16 ": ..."`